### DISCUSSION TITLE
Qual gateway para pagamento usar com React?

### DISCUSSION THREAD

**Discussion by Juuhz** [frontendbr/forum#1478] [Perguntas e Respostas]
Fala Devs, beleza?

Eu estou montando um e-commerce do zero, utilizando NextJS, Node, Styled-Components e Atomic Design.

Eu tenho experiência com checkout do pagseguro, como ele é REST, consigo manipular bem no React/Node, porém a documentação é uma droga e os retornos são em XML ( risos risos ).

Mas estou na dúvida em qual gateway de pagamento utilizar. 

Quais vocês já utilizaram e quais me recomendam ?

Requisitos:

1) API ser REST com retorno Json ( de preferência ).
2) Taxa menores.
3) Dashboard para poder gerenciar as transações. 

Conseguem me ajudar ?

Abraços.

**Comment by rafaelverger** [comment]
Acho que requisitos 1 e 3 são cumpridos por quase todos os possíveis gateways do mercado... Sobre taxas, provavel que no início vc só encontre mais elevadas (já que vc ainda não tem volume de transações pra barganhar)

Sugestões:
No quesito facilidade de implementacao: Pagar.me (3.79% + R$0,50 por sucesso, 30dias pra receber) e Stripe (3.99% + R$0,50 por sucesso, 30dias pra receber) tão bem

Outras opções com alguma facilidade e, as vezes, melhor gerência em assinaturas: Mundipagg ~e Iugu~

Outras considerações:
- O checkout transparente do PagSeguro era **MUITO** zoado, não sei como tá agora.
- Nunca usei o serviço do MercadoPago.
- O serviço da maxiPago! tinha bom preço, implementei e cheguei a usar em produção, mas tiveram um problema (de implementação interna) qd fecharam parceria com a PayPal e daí eles pararam de funcionar e paramos de usar ele (ficamos 1 semana sem conseguir processar nada com eles).

**Reply by rafaelverger** [reply]
**Atualizando aqui as taxas:**
 - Pagar.me: 4,49% + R$0,99
 - Stripe: 3,99% + R$0,39

Com isso o Stripe fica sendo muito superior que o Pagar.me, porém a oferta de bandeiras da Stripe é inferior à da Pagar.me 😢 
 - Pagar.me: Todas as bandeiras (crédito e débito), além de Boleto e **PIX**
 - Stripe: Visa e MasterCard (crédito e débito), além de Boleto e das carteiras virutais Alipay, Apple, Google, Microsoft, Wechat e Click to Pay

**Reply by ftzi** [reply]
Tô usando a stripe agora, e só fui perceber esse pequeno suporte às bandeiras a pouco tempo. Chato demais. Tá um inferno essa questão de gateways.

**Reply by luanrjjj** [reply]
O problema da stripe é a questão do parcelamento. Não é possivel parcelar, o que prejudica bastante.

**Reply by MullerHub** [reply]
> **Atualizando aqui as taxas:**
> 
> * Pagar.me: 4,49% + R$0,99
> * Stripe: 3,99% + R$0,39
> 
> Com isso o Stripe fica sendo muito superior que o Pagar.me, porém a oferta de bandeiras da Stripe é inferior à da Pagar.me 😢
> 
> * Pagar.me: Todas as bandeiras (crédito e débito), além de Boleto e **PIX**
> * Stripe: Visa e MasterCard (crédito e débito), além de Boleto e das carteiras virutais Alipay, Apple, Google, Microsoft, Wechat e Click to Pay

Stripe não tem opção de pagamento via pix?
Já vi monte projetinho de pagamento por pix open-source, será que o stripe não tem mesmo pagamento por pix incluso?

**Reply by luanrjjj** [reply]
Stripe não tem ainda pix no Brasil. Eles prometeram que seria esse ano. Stripe tem alguns detalhes que devem ser levados em conta como a taxa de contestação. Implementei stripe e pagarme recentemente. Pagarme foi bem de boa. Boleto, cartão de crédito e pix é tranquilo. Assinatura é mais complexo

**Comment by hugocarreira** [comment]
https://www.mundipagg.com/

Gateway do grupo Stone, API Rest, Dashboards bem desenvolvidos e informativos, acho que vale a pena dar uma olhada. (Taxação é feita em centavos/transação)

**Reply by rafaelverger** [reply]
Update aqui, MundiPagg agora tb é Pagar.me. A Pagar.me v2 é o resultado da fusão com a Mundi (reorganização da Stone) e daí juntaram o motor de recorrência da Mundi com o motor de pagamentos da Pagarme (a grosso modo).

**Comment by didiraja** [comment]
Pagar.me e Mundipagg, com certeza

**Comment by Juuhz** [comment]
Após escolher bastante e analisar as integrações, a pagar.me me atendeu muito bem! 

Antes de escolher, eu liguei para o suporte e simulei um problema, rapaz, me surpreendeu. Fui super bem atendido e fiquei bem a vontade, eles usam uma linguagem mais informal, parece que eu estava conversando com um amigo meu. 

Ah, e foram super solícitos!

Quem quiser ver a integração funcionando, só acessar:

[http://cardesivo.com.br](http://cardesivo.com.br)

---

Qualquer dúvida sobre integração eu posso ajudar.

**Reply by monecchi** [reply]
A finalização da compra pelo chrome tá bugada... Depois vou comprar meus cardesivos! Valeu 👍

**Reply by tainarodriguesc** [reply]
Quanto tempo levou para conseguir integrar com a pagarme?

**Comment by yuritoledo** [comment]
@Juuhz Sabe alguma plataforma que integre bem com react native?
Estou apanhando feio :(

**Comment by faelribeiro22** [comment]
Não sei se ajuda em algo, mas tem esse aqui:
https://stripe.com/en-br/payments/elements

**Comment by yuritoledo** [comment]
@faelribeiro22 pelo que vi o stripe, não tem "suporte" para contas bancária (eu acho) do brasil

**Comment by Juuhz** [comment]
Fala @yuritoledo, beleza? 

Você pode utilizar o próprio Pagar.me, eles tem 2 tipos de soluções:

- Checkout em lightbox.
- Checkout em REST.

A diferença entre os 2 é que: Com o lightbox ele já cria os campos de: cartão, nome, numero, etc... e eu não preciso me preocupar com validações.  

Mas como você está usando o React Native, "acredito" que não deva funcionar o lightbox ( mas vale o teste, depois me conta ). 

Mas você pode usar o REST + qualquer módulo de checkout em React Native, exemplo:
https://www.npmjs.com/package/react-native-checkout

Resumindo:
Você cria uma tela de checkout, com todos os campos: nome, email, cartão de crédito, etc... Depois compila essas informações e envia para o pagar.me, para criar a transação. 

A doc deles é bem boa: 
https://docs.pagar.me/docs/overview-transacao

Para se registrar e já gerar uma API de sandbox:
https://beta.dashboard.pagar.me/#/account/signup

Qualquer dúvida estou à disposição.

**Comment by yuritoledo** [comment]
@Juuhz opa! Obrigado pelo tempo e atenção :)

Ontem cheguei exatamente nessas mesmas opções que comentou.
Tentei rodar aquele modal de checkout (deve ser esse lightbox que você comenta) em uma webview, infelizmente sem sucesso. 
Baixei o js e tentei fazer umas edições, mas também não deram certo.

Acabei optando pela segunda opção, faço um POST para a API e pego os dados através de um componente de checkout, parecido com esse que mandou. Foi esse aqui: https://github.com/sbycrosz/react-native-credit-card-input 

Tem alguns errinhos bobos e é meio simples, vou fazer umas PRs depois pra dar uma ajudada

Em suma, é isso :grin:

**Comment by Juuhz** [comment]
Boa @yuritoledo, ta no caminho certo, jaja irá finalizar.

**Comment by romulosardinha** [comment]
Próximo passo de uma aplicação minha vai ser o meio de pagamento tbm. vou usar essa dica de vocês do pagar.me. vlw ai

**Comment by yuritoledo** [comment]
@rmlSardinha Qualquer coisa, estamos aí :)

**Comment by lucaashrq** [comment]
Fala @Juuhz, beleza?

Fugindo um pouco do assunto do tópico, estou desenvolvendo uma aplicação em NextJS e estou com problemas para realizar o deploy.

Como você fez? Qual ambiente utilizou?
Se puder dar esse força, tô apanhando bastante hehehe

**Comment by spacedog4** [comment]
@rafaelverger eu sempre quis utilizar o stripe, mas ele parece dar suporte ao mundo todo menos ao Brasil, como conseguiu utilizar? ao acessar o site ele pede para eu informar meu email que quando estiver disponível eles me avisam

**Comment by rafaelverger** [comment]
@spacedog4 eu pedi pra entrar no beta deles, salvo engano eles iriam iniciar a operação Brasil esse ano!

**Comment by Juuhz** [comment]
> Fala @Juuhz, beleza?
> 
> Fugindo um pouco do assunto do tópico, estou desenvolvendo uma aplicação em NextJS e estou com problemas para realizar o deploy.
> 
> Como você fez? Qual ambiente utilizou?
> Se puder dar esse força, tô apanhando bastante hehehe

O meu deploy é manual e bem básico (um dia eu coloco no jenkins), mas funciona assim:

1) Entro no servidor via ssh.
2) Entro na pasta do projeto.
3) Dou pull no repositório.
4) Rodo o build.
5) Reinicio o nginx.

gg izi.

**Comment by Juuhz** [comment]
Galera, 

Após quase 1 ano utilizando o Pagar.me como solução de gateway para meu e-commerce feito em NextJs, tenho algumas impressões.

**Pontos positivos:**
- A solução é ótima do lado do desenvolvimento, API bem explicada e requisições rápidas.
- Suporte é ótimo também, bem humanizado, não é aqueles robôs humanos que só dão CTRL+V.
- Dashboard (beta) bem organizado e prático, mas ainda possui alguns bugs, mas como é beta, não vou criticar, me atende bem.
- O Checkout transparente deles é maravilhoso, meus clientes sempre elogiam, é fácil e rápido.

**Pontos negativos:**
- As taxas não são muito competitivas:
![image](https://user-images.githubusercontent.com/7996030/80834502-36101d80-8bc7-11ea-9aa8-6ad804ba51c3.png)
- Não tem integração com o conta-azul, tive que desenvolver por conta própria.
- Não curti o antifraude deles, pago relativamente caro R$ 0,70 por transação e não me passa uma confiabilidade boa.

-----

Agora estou implementando o PicPay que está com umas taxas bem interessantes.

https://ecommerce.picpay.com/

**Reply by duccini** [reply]
> Galera,
> 
> Após quase 1 ano utilizando o Pagar.me como solução de gateway para meu e-commerce feito em NextJs, tenho algumas impressões.
> 
> **Pontos positivos:**
> 
> * A solução é ótima do lado do desenvolvimento, API bem explicada e requisições rápidas.
> * Suporte é ótimo também, bem humanizado, não é aqueles robôs humanos que só dão CTRL+V.
> * Dashboard (beta) bem organizado e prático, mas ainda possui alguns bugs, mas como é beta, não vou criticar, me atende bem.
> * O Checkout transparente deles é maravilhoso, meus clientes sempre elogiam, é fácil e rápido.
> 
> **Pontos negativos:**
> 
> * As taxas não são muito competitivas:
>   ![image](https://user-images.githubusercontent.com/7996030/80834502-36101d80-8bc7-11ea-9aa8-6ad804ba51c3.png)
> * Não tem integração com o conta-azul, tive que desenvolver por conta própria.
> * Não curti o antifraude deles, pago relativamente caro R$ 0,70 por transação e não me passa uma confiabilidade boa.
> 
> Agora estou implementando o PicPay que está com umas taxas bem interessantes.
> 
> https://ecommerce.picpay.com/

Mas não limita muito? Muitos clientes não tem cadastro no picpay e nem querem fazê-lo.

**Comment by monecchi** [comment]
Cara bem legal as sugestões aqui, mas eu vi um serviço em um site e-commerce que vendia produtos de vestuário, só que não lembro mais o nome, vou precisar encontrar de novo, achei super intuitivo o checkout deles, com etapas bem resumidas, e um design de interface moderno, inteligente e bonito...

Na hora, pesquisei rapidamente e descobri que era uma solução completa, com bibliotecas (framework) das etapas do checkout já prontos pra usar com React... O serviço de gateway de pagamentos deles parecia bem robusto e responsivo, tipo o Stripe, incluindo uma gama boa de meios de pagamentos aceitos, com suporte à API REST de forma isolada (onde as implementações da captura e do checkout ficavam por conta do desenvolvedor), e também tinha uma implementação drop-in pra web e apps nativos pelo framework deles mesmos...

Se alguém passar por esse issue e souber sobre esse serviço deixa o link aqui por favor, até lá, vou tentar encontrar nos meus favoritos, que são centenas e centenas...

**Comment by Marcheto88** [comment]
> Cara bem legal as sugestões aqui, mas eu vi um serviço em um site e-commerce que vendia produtos de vestuário, só que não lembro mais o nome, vou precisar encontrar de novo, achei super intuitivo o checkout deles, com etapas bem resumidas, e um design de interface moderno, inteligente e bonito...

Cara, se encontrar, poderia postar aqui? Estou procurando algo exatamente como vc descreveu..

**Comment by Marcheto88** [comment]
@Juuhz blz? Cai aqui procurando um gateway pra usar e curti bastante teu site. Fui até o checkout e gostaria de saber se vc tem algum medidor de quantas pessoas deixam o carrinho e não finalizam a compra? Eu achei bom o checkout transparente, mas me pareceu que um usuário simples possa ficar com medo de inserir os dados ali e largar o carrinho.

Estou começando no react e pretendo colocar um ecommerce de um produto só o mais rapido possível no ar, bem no estilo do teu. Vamos ver se vai sair hehe

**Comment by lucashfreitas** [comment]
Opa @Juuhz ,

Venho utilizando o stripe por cerca de 2 anos e de todos os payments gateways que utilizei (ja utilizei world pay, 2checkout, pagseguro).  Não vi nenhum chegar perto do stripe no quesito a tecnologia, facilidade de integracao, monitoramento e suporte ao usuàrio. 

Me parece que o stripe ja esta disponivel no Brazil, nao sei se as taxas sao vantajosas igual as outras opcoes, mas vale apena dar uma conferida. 

Eu recebi o convite e ja tenho uma conta ativa no stripe no BR e ta funcionando legal 👍

**Comment by dev-doug-k4** [comment]
Estou utilizando Stripe em um app desenvolvido com react-native. Funciona muito bem (ao menos no ambiente de teste rsr), a documentação deles é a mais completa que encontrei, com vários exemplos de integração e existem muitos tutoriais de terceiros fáceis de encontrar. O problema é que minha aplicação é um marketplace e estou na fila para liberarem uma conta Connect Custom (no momento só tem disponível a Connect Standard no Brasil) e pela demora estou cogitando utilizar outra solução. Acabei encontrando a Iugu. Porém, a experiencia inicial não foi das melhores, mandei alguns e-mails com algumas dúvidas e até hoje não me responderam. Alguém aí já utilizou Iugu?

**Comment by ftzi** [comment]
Perdão por comentar numa discussão antiga, mas tenho a seguinte questão: No meu App, parte dos pagamentos dos clientes será transferida para as empresas prestadoras do serviço. O Stripe demora 30 dias para fazer a transferência do pagamento para a conta bancária de destino. É muito tempo, não seria agradável para as prestadoras.

Têm soluções melhores, nesse quesito de tempo?

**Reply by rafaelverger** [reply]
Acho que tem vários gateways que fazem o share de marketplace e quase todos tem a possibilidade de pagamento em 30d por padrão, mas vc pode configurar um adiantamento (a taxa vai ser maior no custo da transação).

Pagar.me e Iugu, por exemplo, possuem o "split de marketplace" (parte da grana cai direto no fornecedor do serviço, parte vai pra plataforma) e ambos possuem o recurso de "adiantamento", que é exatamente o adiantar os recebíveis que vc teria em 30dias pra 2 dias, por exemplo

**Reply by iltonbarbosa** [reply]
Olá SrBrahma, tudo blz?
Conseguiu implementar um outro gatway de pagamento ou ainda está utilizando o Stripe?

**Reply by ftzi** [reply]
Rapaz, isso me deu um estresse fudido no tempo que fiquei nesse projeto.

Stripe eh a melhor opcao mesmo (agora suporta Pix inclusive), os outros gateways brasileiros sao bem limitados no geral, exceto caso tu tenha um projeto ja muito bem consolidado e que lide com um fluxo de dinheiro la na casa das centenas de milhares ou milhoes.
